### PR TITLE
Resettables

### PR DESCRIPTION
**Pull request by tkstanczak**
DRY for resettable caches in storage and state providers. Consistent approach to clear vs new on arrays, dictionaries and hashsets in the store namespace.

Also, PC is int here instead of UInt256.

### CODE REVIEW

**Comment by LukaszRozmej** [comment]
Is PC as int a good idea? maybe uint or ulong would be better?

File: src/Nethermind/Nethermind.Store/ResettableDictionary.cs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Nethermind.Store
+{
+    public class ResettableDictionary<T1, T2> : IDictionary<T1, T2>
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Store/ResettableDictionary.cs]: standard nomenclature is TKey, TValue

File: src/Nethermind/Nethermind.Store/ResettableDictionary.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            get => _wrapped[key];
+            set => _wrapped[key] = value;
+        }
+
+        public ICollection<T1> Keys => _wrapped.Keys;
+        public ICollection<T2> Values => _wrapped.Values;
+
+        public void Reset()
+        {
+            if (_wrapped.Count == 0)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Store/ResettableDictionary.cs]: Looks like code duplication between Dictionary and HashSet but not sure if worth extracting.

**Comment by tkstanczak** [comment]
> Is PC as int a good idea? maybe uint or ulong would be better?

max code size on mainnet is
private const int MaxCodeSize = 24576;

PC is a position in the byte code array

we would do with short but int is preferred since it is one variable only and int will have best perf

File: src/Nethermind/Nethermind.Store/ResettableDictionary.cs (hunk 1 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Store/ResettableDictionary.cs]: will fix

File: src/Nethermind/Nethermind.Store/ResettableDictionary.cs (hunk 2 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Store/ResettableDictionary.cs]: I made it generic at the beginning and ended up with something like:

Resettable<Dictionary<StorageAddress, Stack<int>>, KeyValuePair<StorageAddress, Stack<int>>>

In the end, copy-paste was much better.

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Store/ResettableDictionary.cs]: Having ResettableDictionary<TKey, TValue> : Resettable<IDictionary<Tkey, TValue>, KeyValuePair<Tkey TValue>>

and moving implementation to abstract Resettable, while end-user would use ResettableDictionary sounds reasonable.

**Pull request closed without merging**